include/pmacc/simulationControl/SimulationHelper.hpp

```diff
@@ -401,41 +404,97 @@ namespace pmacc
+                 * with blocking collective operations. Each MPI rank can be in different time steps and phases
+                 * of the simulation, therefore we can not assume that all MPI ranks received the signal in the same
+                 * time step
+                 */
+
                 if(output)
                     std::cout << "SIGNAL: received." << std::endl;
 
-                // avoid deadlocks with collective MPI calls
-                Environment<>::get().Manager().waitForAllTasks();
                 // wait for possible more signals
                 std::this_thread::sleep_for(std::chrono::milliseconds(1000u));
 
-                uint32_t maxTimestep = 0u;
+                /* After a signal is received we need to perform one more time step to avoid dead-locks if a
+                 * simulation phase is using blocking MPI collectives. After the additional step we know that
+                 * all MPI ranks participated in MPI_Iallreduce.
+                 */
+                handleSignalAtStep = currentStep + 1;
 
                 // find largest time step of all MPI ranks
-                MPI_CHECK(MPI_Allreduce(
-                    &currentStep,
-                    &maxTimestep,
+                MPI_CHECK(MPI_Iallreduce(
+                    &handleSignalAtStep,
+                    &signalMaxTimestep,
                     1,
                     MPI_UINT32_T,
                     MPI_MAX,
-                    Environment<DIM>::get().GridController().getCommunicator().getMPIComm()));
+                    Environment<DIM>::get().GridController().getCommunicator().getMPISignalComm(),
+                    &signalMPI));
 
                 if(signal::createCheckpoint())
                 {
                     if(output)
-                        std::cout << "SIGNAL: active checkpointing for step " << maxTimestep << std::endl;
-                    // add a new checkpoint
-                    seqCheckpointPeriod.push_back(pluginSystem::TimeSlice(maxTimestep, maxTimestep));
+                        std::cout << "SIGNAL: Received at step " << currentStep << ". Schedule checkpointing. "
+                                  << std::endl;
+                    signalCreateCheckpoint = true;
                 }
                 if(signal::stopSimulation())
                 {
                     if(output)
-                        std::cout << "SIGNAL: stop simulation at step " << maxTimestep << std::endl;
-                    Environment<>::get().SimulationDescription().setRunSteps(maxTimestep);
+                        std::cout << "SIGNAL: Received at step " << currentStep << ". Schedule shutdown." << std::endl;
+                    signalStopSimulation = true;
+                }
+            }
+            /* We will never handle a signal at step zero.
+             * If we received a signal handleSignalAtStep will be set to currentStep + 1 (see above)
+             */
```

Comment:
I am confused, doesn't this comment contradict the first comment of this function? I mean the handling on step 0 part

Reply:
I will rephrase it and will separate `handleSignals` from the step zero. We have two phases here. One is seeing that a signal arrives and start the process to find the largest simulation step between MPI ranks and the second phase where we can translate the signal one simulation step after the signal arrived into an explicit action.